### PR TITLE
feat: add command for showing AST and named AST

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR tries to add in a `showAst` and `showAstNamed` command, which will scan the currently opened file, and output the pretty-printed AST.

## Why:
This would be really helpful for people developing Semgrep, or people who are writing rules, I reckon.

## How:
I added in the commands, similarly to the other ones included. I followed `rust-analyzer`, for the most part. I didn't touch anything related to actually interpreting the output that is generated from the command, especially for presenting it within the UI, but I have vague hopes that the output will end up somewhere.

## Test plan:
See comment below!

### Security

- [ ] Change has security implications (if so, ping the security team)

### CODE REVIEW

**Review by ajbt200128 — CHANGES_REQUESTED** [review]
So the main thing missing here is actually showing the output somewhere. I think there are 3 main ways of doing this I've seen (in order of amount of work):

1. send over the AST as just formatted text (I think you're doing this now), open a new file to the side of the current and just display it
2. Convert the AST to some sort of JSON (not sure what this would look like) and display it in a [tree view ](https://code.visualstudio.com/api/extension-guides/tree-view), here's what I did for the search demo [branch](https://github.com/returntocorp/semgrep-vscode/tree/austin/search)
3. Do what ocamllabs does, which is custom react [link](https://github.com/ocamllabs/vscode-ocaml-platform/tree/master/astexplorer)

My preference is 2, though I'd be ok with 1. What ocamllabs does is... kind of weird, I don't think it's recommended by Microsoft, and I think the tree view has the same functionality.

To get the highlight text -> show ast functionality you mentioned, you're almost there, you just would add a context menu option [ref](https://code.visualstudio.com/api/ux-guidelines/context-menus) that gets the selection range and then pass that range to the showAst commands (VSCode makes it easy to call commands internally once they're registered), and on the ocaml side, read the file and trim it to that range.

File: src/commands.ts [hunk 1]
```diff
@@ -32,6 +33,22 @@ export function registerCommands(
     await client.sendNotification(scanWorkspace, { full: true });
   });
 
+  vscode.commands.registerCommand("semgrep.showAst", async () => {
+    // supposedly the right thing to do, according to
+    // https://github.com/Microsoft/vscode/issues/36426
+    await client.sendRequest(showAst, {
```
Comment (ajbt200128) [inline src/commands.ts:39]: `sendRequest` returns the result of the request, so in your case the AST. this is where you'd want to call some function to show it in the UI

**Comment by brandonspark** [comment]
@ajbt200128 I added some UI stuff, according to these two sources:
https://github.com/microsoft/vscode-extension-samples/blob/main/contentprovider-sample/src/extension.ts
https://code.visualstudio.com/api/extension-guides/virtual-documents

I wasn't sure of how to turn the Generic AST into a proper JSON value, so right now I'm just going for the text document method. Down to change this later if I figure that out + once I get my bearings on this VS Code thing a bit more.

**Comment by brandonspark** [comment]
I also... still don't know how to test this locally. Would love to sync about it, if you have time to show me.

File: src/commands.ts [hunk 2]
```diff
@@ -35,6 +37,34 @@ export function registerCommands(
     await client.sendNotification(scanWorkspace, { full: true });
   });
 
+  vscode.commands.registerCommand("semgrep.showAstNamed", async () => {
+    const ast_text = await client.sendRequest(showAst, {
+      named: true,
+      uri: vscode.window.activeTextEditor.document.uri.fsPath,
+    });
+    env.documentView.setText(ast_text);
+    const uri = encodeUri(vscode.editor.document.uri);
```
Comment (ajbt200128) [inline src/commands.ts:46]:
```suggestion
    const uri = encodeUri(vscode.window.activeEditor.document.uri);
```

File: src/commands.ts [hunk 3]
```diff
@@ -35,6 +37,34 @@ export function registerCommands(
...
+      );
+  });
+
+  vscode.commands.registerCommand("semgrep.showAst", async () => {
+    const ast_text = await client.sendRequest(showAst, {
+      named: false,
+      uri: vscode.window.activeTextEditor.document.uri.fsPath,
+    });
+    env.documentView.setText(ast_text);
+    const uri = encodeUri(vscode.editor.document.uri);
```
Comment (ajbt200128) [inline src/commands.ts:60]:
```suggestion
    const uri = encodeUri(vscode.window.activeEditor.document.uri);
```

**Review by ajbt200128 — COMMENTED** [review]
I think for the text ast approach using this virtual documents approach makes sense! Overall looks good, we should probably document these commands in the readme

**Comment by brandonspark** [comment]
https://github.com/returntocorp/semgrep-vscode/assets/49291449/4e2f683a-701a-4759-b213-71397896f9db

File: package.json [hunk 4]
```diff
@@ -59,6 +59,14 @@
         "command": "semgrep.refreshRules",
         "title": "Semgrep: Update rules"
       },
+      {
+        "command": "semgrep.showAst",
+        "title": "Semgrep: Show Generic AST"
+      },
+      {
+        "command": "semgrep.showAstNamed",
```
Comment (ajbt200128) [inline package.json:67]: Can you add a `when` clause and check the minor version here? I'm guessing it'll prolly be minor version 33. See the search command below for how to do that

File: src/commands.ts [hunk 5]
```diff
@@ -10,11 +10,30 @@ import {
...
+  uri: vscode.Uri,
+  content: string
+): Promise<void> {
+  const doc = await vscode.workspace.openTextDocument(uri);
+  const edit = new vscode.WorkspaceEdit();
+  edit.replace(uri, new vscode.Range(0, 0, doc.lineCount, 0), content);
+  vscode.workspace.applyEdit(edit);
+  vscode.window.showTextDocument(
+    doc,
+    vscode.window.activeTextEditor!.viewColumn! + 1
```
Comment (ajbt200128) [inline src/commands.ts:33]: Nit: I would add a check here for the active text editor. I know it seems unlikely but a user /may/ run this command without a text editor focused, i.e. they clicked on the sidebar or they're on the welcome page

File: src/commands.ts [hunk 6]
```diff
@@ -32,6 +51,28 @@ export function registerCommands(env: Environment): void {
...
+    replaceAndOpenUriContent(uri, ast_text);
+  });
+
+  vscode.commands.registerCommand("semgrep.showAst", async () => {
+    const ast_text = await env.client!.sendRequest(showAst, {
+      named: false,
+      uri: vscode.window.activeTextEditor?.document.uri.fsPath,
+    });
+    const uri = encodeUri(
+      vscode.window.activeTextEditor?.document.uri || vscode.Uri.parse("")
```
Comment (ajbt200128) [inline src/commands.ts:71]: Same nit here as above, I'd add a check for active editor

**Review by ajbt200128 — APPROVED** [review]
Looks good, added a few nits.

**Comment by brandonspark** [comment]
@ajbt200128 finally got around to this. I verified that no crash happens when running `showAst` with no active window.

**Pull request merged**